README.rst

===============
pytest-datatest
===============

..
    Project badges for quick reference:

|TravisCI_status| |mit_license|


The development version of `datatest`_'s pytest integration.

Typically, only developers working on datatest's pytest plugin
should install this package. Other users are encouraged to install
just ``datatest``—this plugin comes bundled with it.


Requirements
============

* `pytest`_ (3.3 or newer)
* `datatest`_ (0.9.0 or newer)


Installation
============

For Developers
--------------

Clone the git repositories and use `pip`_ to perform an
editable install from the cloned project directory::

    git clone git@github.com:shawnbrown/datatest.git
    git clone git@github.com:shawnbrown/pytest-datatest.git
    pip install --editable ./datatest
    pip install --no-deps --editable ./pytest-datatest


For Users
---------

Don't install this package—install `datatest`_ instead::

    pip install datatest


For Users With Special Requirements
-----------------------------------

If you aren't a datatest or plugin developer but instead need
bug-fixes or features that are not yet available from the main
datatest project, you can install "pytest-datatest" via pip
from `PyPI`_::

    pip install datatest
    pip install pytest-datatest

And if the version is PyPI isn't new enough for your needs, you can
install the plugin directly from the live GitHub repository (make
sure to check that the build is "passing" before doing this)::

    pip install datatest
    pip install --upgrade https://github.com/shawnbrown/pytest-datatest/archive/master.zip


Usage
=====

When installed, this development version automatically overrides
datatest's bundled pytest integration. If you want to revert back
to the bundled plugin, simply uninstall this version.

To disable the development version (and temporarily enable the
bundled version) use::

    pytest -p no:datatest_devel


To disable both the development *and* bundled versions use::

    pytest -p no:datatest_devel -p no:datatest


Features
========

* Provides a 'mandatory' marker to support incremental testing.
* Provides an '--ignore-mandatory' command line option to override
  the default 'mandatory' behavior.
* Strips the leading "E   " prefix from ``ValidationError`` *differences*
  to help users more easily repurpose parts of the error  message for
  ``allowed.specific(...)`` definitions.


License
=======

Distributed under the terms of the `MIT`_ license, "pytest-datatest" is
free and open source software


Issues
======

If you encounter any problems, please `file an issue`_ along with a
detailed description.


.. |TravisCI_status| image:: https://travis-ci.org/shawnbrown/pytest-datatest.svg?branch=master
    :target: https://travis-ci.org/shawnbrown/pytest-datatest
    :alt: Travis CI Build Status
.. |AppVeyor_status| image:: https://ci.appveyor.com/api/projects/status/github/shawnbrown/pytest-datatest?branch=master
    :target: https://ci.appveyor.com/project/shawnbrown/pytest-datatest/branch/master
    :alt: AppVeyor Build Status
.. |devstatus| image:: https://img.shields.io/pypi/status/pytest-datatest.svg
    :target: https://pypi.python.org/pypi/pytest-datatest
    :alt: Development Status
.. |mit_license| image:: https://img.shields.io/badge/license-MIT-blue.svg
    :target: http://opensource.org/licenses/MIT
    :alt: MIT License
.. |pyversions| image:: https://img.shields.io/pypi/pyversions/pytest-datatest.svg
    :target: https://pypi.python.org/pypi/pytest-datatest#supported-versions
    :alt: Supported Python Versions
.. _`datatest`: https://pypi.python.org/pypi/datatest
.. _`file an issue`: https://github.com/shawnbrown/pytest-datatest/issues
.. _`MIT`: http://opensource.org/licenses/MIT
.. _`pip`: https://pypi.python.org/pypi/pip/
.. _`PyPI`: https://pypi.python.org/pypi
.. _`pytest`: https://pypi.python.org/pypi/pytest
.. _`tox`: https://tox.readthedocs.io/en/latest/


CHANGELOG

=========================
pytest_datatest Changelog
=========================


2020-10-14 (0.1.4)
==================

* Fixed bug where ValidationError formatting was crashing xdist workers.
* Changed internals to make using pytest's non-public API more robust.
* Added testing support for Python 3.8 and 3.9.


2019-01-28 (0.1.3)
==================

* This update adds testing and support for latest versions of Pytest and
  Python (now tested using Pytest 3.3 to 4.1 and Python 2.7 to 3.7).
* Changed handling for 'mandatory' marker to support older and newer Pytest
  versions.


2018-06-21 (0.1.2)
==================

* Added handling for a 'mandatory' marker to support incremental testing
  (stops session early when a mandatory test fails).
* Added '--ignore-mandatory' option to continue tests even when a mandatory
  test fails.
* Changed the bundled-version-notice from a *warning* into a report section.
* Added pytest hook to properly deactivate the bundled plugin when using the
  development version.


2018-04-07 (0.1.1)
==================

* First public release.
* Added formal development-version notice (see module docstring and README).
  This plugin is primarily intended for developers while typical users should
  use the copy that is bundled with ``datatest``.
* Added a warning if the plugin that's bundled with ``datatest`` is newer.
* Added error formatting when pytest runs existing unittest-style tests.


2017-11-25 (0.1.0)
==================

* Initial version for testing (never officially released).
* Added error message formatting to improve ``datatest`` workflow.


Changelog Guidelines
====================

* Begin each section with the date followed by the version number in
  parenthesis. Use the following format: "YYYY-MM-DD (x.y.z)".
* The initial bullet-point may provide a one-line description of the release.
* Following bullet-points should begin with "Added", "Changed", "Fixed", or
  "Removed" when describing the notable changes.
* Limit lines to 80 character width.


tests/test_datatest.py
# -*- coding: utf-8 -*-
import pytest

from pytest_datatest import _find_validationerror_position
from pytest_datatest import _formatted_lines_generator


class DummyTerminalWriter(object):
    """Helper class for testing calls to line() and write() methods."""
    def __init__(self):
        self.all_lines = []

    def line(self, line, **kwds):
        self.all_lines.append((line, kwds))

    def write(self, line, **kwds):
        self.all_lines.append((line, kwds))


class TestFindValidationErrorPosition(object):
    def test_fully_qualified_class_name(self):
        """First error line uses 'datatest.ValidationError'."""
        lines = [
            '    def test_mydata(self):',
            '>       validate(1, 2)',
            'E       datatest.ValidationError: invalid data (1 difference): [',
            'E           Deviation(-1, 2),',
            'E       ]',
            '',
            'test_script.py:42: ValidationError',
        ]
        position = _find_validationerror_position(lines)
        assert position == 2, 'expects start at index 2'

    def test_unqualified_class_name(self):
        """First error line uses 'ValidationError'."""
        lines = [
            '    def test_mydata(self):',
            '>       validate(1, 2)',
            'E       ValidationError: invalid data (1 difference): [',
            'E           Deviation(-1, 2),',
            'E       ]',
            '',
            'test_script.py:42: ValidationError',
        ]
        position = _find_validationerror_position(lines)
        assert position == 2, 'expects start at index 2'

    def test_no_validationerror_found(self):
        """Check AssertionError, not ValidationError."""
        lines = [
            "    def test_foobar():",
            ">       assert 'foo' == 'bar'",
            "E       AssertionError: assert 'foo' == 'bar'",
            "E         - bar",
            "E         + foo",
            "",
            "test_script.py:42: AssertionError",
        ]
        position = _find_validationerror_position(lines)
        assert position == -1, 'not found, should be -1'

    def test_nested_validationerror_string(self):
        """Check that there are no false positives for lines that look
        like ValidationErrors but occur after the first line with a
        fail_marker.
        """
        lines = [
            r"    def test_bar():",
            r">       assert 'foo' == '''",
            r"    \b\b\b\bdatatest.ValidationError: invalid data (1 difference): [",
            r"    '''",
            r"E       AssertionError: assert 'foo' == '\n\x08\x08\x...ference): [\n'",
            r"E         + foo",
            r"E         - ",
            r"E       datatest.ValidationError: invalid data (1 difference): [",
            r"",
            r"test_script.py:42: AssertionError",
        ]
        position = _find_validationerror_position(lines)
        assert position == -1, 'not found, should be -1'


class TestFormattedLinesGenerator(object):
    def test_formatting(self):
        lines = [
            '    def test_mydata(self):',
            '>       validate(1, 2)',
            'E       datatest.ValidationError: invalid data (1 difference): [',
            'E           Deviation(-1, 2),',
            'E       ]',
            '',
            'test_script.py:42: ValidationError',
        ]
        position = _find_validationerror_position(lines)
        formatted = _formatted_lines_generator(lines, position)

        expected = [
            '    def test_mydata(self):',
            '>       validate(1, 2)',
            'E       ValidationError: invalid data (1 difference): [',
            '            Deviation(-1, 2),',  # <- No "E" prefix!
            '        ]',                      # <- No "E" prefix!
            '',
            'test_script.py:42: ValidationError',
        ]
        assert list(formatted) == expected

    def test_nongreedy_matching(self):
        """Should stop removing `fail_marker` characters after the
        first line without a matching `fail_marker`.
        """
        lines = [
            '    def test_mydata(self):',
            '>       validate(1, 2)',
            'E       datatest.ValidationError: invalid data (1 difference): [',
            'E           Deviation(-1, 2),',
            'E       ]',
            '',
            'Etest_script.py:42: ValidationError',  # <- Starts with fail_marker match!
        ]
        position = _find_validationerror_position(lines)
        formatted = _formatted_lines_generator(lines, position)

        expected = [
            '    def test_mydata(self):',
            '>       validate(1, 2)',
            'E       ValidationError: invalid data (1 difference): [',
            '            Deviation(-1, 2),',        # <- No "E" prefix!
            '        ]',                            # <- No "E" prefix!
            '',
            'Etest_script.py:42: ValidationError',  # <- Should be unchanged!
        ]
        # Comparing the last two lines is enough.
        assert list(formatted)[-2:] == expected[-2:]


class TestReprEntryLinesFormatting(object):
    """The pytest-datatest plugin uses the pytest_runtest_logreport()
    hook to format the ReprEntry.lines for datatest.ValidationError
    failures.

    This will have the effect of removing the "E" prefix from the
    difference rows shown in error messages.
    """
    def test_simple_failure(self, testdir):
        """Test simple failure case."""

        testdir.makepyfile('''
            from datatest import ValidationError
            from datatest import Invalid

            def test_validation():
                raise ValidationError([Invalid('a')], 'invalid data')
        ''')
        result = testdir.runpytest('-v')

        result.stdout.fnmatch_lines([
            "E       ValidationError: invalid data (1 difference): [",
            "            Invalid('a'),",  # <- No "E" prefix!
            "        ]",                  # <- No "E" prefix!
            "",
        ])

    def test_wrapped_failure(self, testdir):
        """Should also replace ReprEntry objects in wrapped functions too."""

        testdir.makepyfile('''
            from datatest import ValidationError
            from datatest import Invalid

            def wrapped():
                raise ValidationError([Invalid('a')], 'invalid data')

            def test_wrapped_call():
                wrapped()
        ''')
        result = testdir.runpytest('-v')

        result.stdout.fnmatch_lines([
            "E       ValidationError: invalid data (1 difference): [",
            "            Invalid('a'),",  # <- No "E" prefix!
            "        ]",                  # <- No "E" prefix!
            "",
        ])

    def test_with_unittester(self, testdir):
        """Test ReprEntry replacement with unittest-style tests."""

        testdir.makepyfile('''
            from datatest import DataTestCase
            from datatest import ValidationError
            from datatest import Invalid

            class TestValidation(DataTestCase):
                def test_validation(self):
                    raise ValidationError([Invalid('a')], 'invalid data')
        ''')
        result = testdir.runpytest('-v')

        result.stdout.fnmatch_lines([
            "E       ValidationError: invalid data (1 difference): [",
            "            Invalid('a'),",  # <- No "E" prefix!
            "        ]",                  # <- No "E" prefix!
            "",
        ])


class TestTruncation(object):
    @pytest.fixture(autouse=True)
    def long_error(self, testdir):
        testdir.makepyfile('''
            from datatest import ValidationError
            from datatest import Invalid

            def test_validation():
                raise ValidationError(
                    [Invalid(x) for x in range(10)],
                    description='invalid data',
                )
        ''')

    def test_default_truncation(self, testdir):
        result = testdir.runpytest('-v')
        result.stdout.fnmatch_lines([
            "E       ValidationError: invalid data (10 differences): [",
            "            Invalid(0),",  # <- No "E" prefix!
            "            Invalid(1),",  # <- No "E" prefix!
            "            Invalid(2),",  # <- No "E" prefix!
            "            Invalid(3),",  # <- No "E" prefix!
            "            Invalid(4),",  # <- No "E" prefix!
            "            Invalid(5),",  # <- No "E" prefix!
            "            Invalid(6),",  # <- No "E" prefix!
            "            Invalid(7),",  # <- No "E" prefix!
            "            ...",          # <- No "E" prefix!
            "        ",
            "        ...Full output truncated, use '-vv' to show",
            "",
        ])

    def test_increased_verbosity(self, testdir):
        result = testdir.runpytest('-vv')
        result.stdout.fnmatch_lines([
            "E       ValidationError: invalid data (10 differences): [",
            "            Invalid(0),",  # <- No "E" prefix!
            "            Invalid(1),",  # <- No "E" prefix!
            "            Invalid(2),",  # <- No "E" prefix!
            "            Invalid(3),",  # <- No "E" prefix!
            "            Invalid(4),",  # <- No "E" prefix!
            "            Invalid(5),",  # <- No "E" prefix!
            "            Invalid(6),",  # <- No "E" prefix!
            "            Invalid(7),",  # <- No "E" prefix!
            "            Invalid(8),",  # <- No "E" prefix!
            "            Invalid(9),",  # <- No "E" prefix!
            "        ]",                # <- No "E" prefix!
            "",
        ])


class TestXdistPlugin(object):
    """Check behavior using the `pytest-xdist` plugin. Communication
    between xdist workers and pytest are managed with JSON messages.
    Unlike non-xdist test sessions, results are processed using the
    _pytest.reports._report_kwargs_from_json() function.
    """

    @pytest.fixture
    def passing_case(self, testdir):
        testdir.makepyfile('''
            from datatest import validate

            def test_one():
                validate('foo', 'foo')

            def test_two():
                validate('bar', 'bar')
        ''')
        return testdir

    @pytest.fixture
    def failing_case(self, testdir):
        testdir.makepyfile('''
            from datatest import validate

            def test_one():
                validate('foo', 'baz')

            def test_two():
                validate('bar', 'baz')
        ''')
        return testdir

    def test_passing(self, passing_case):
        """Run passing cases with xdist option '-n 1'."""
        result = passing_case.runpytest('-n', '1')
        result.assert_outcomes(passed=2, failed=0)

    def test_failing(self, failing_case):
        """Run failing cases with xdist option '-n 1'."""
        result = failing_case.runpytest('-n', '1')
        result.assert_outcomes(passed=0, failed=2)


class TestMandatoryMarker(object):
    def test_marker_registered(self, testdir):
        result = testdir.runpytest('--markers')
        result.stdout.fnmatch_lines([
            '@pytest.mark.mandatory:*',
        ])

    @pytest.fixture
    def passing_cases(self, testdir):
        testdir.makepyfile('''
            import pytest

            @pytest.mark.mandatory
            def test_first():
                pass

            def test_second():
                pass
        ''')
        return testdir

    @pytest.fixture
    def failing_cases(self, testdir):
        testdir.makepyfile('''
            import pytest

            @pytest.mark.mandatory
            def test_first():
                raise Exception()

            def test_second():
                raise Exception()
        ''')
        return testdir

    def test_session_finished(self, passing_cases):
        result = passing_cases.runpytest()
        result.assert_outcomes(passed=2, failed=0)

    def test_session_stopped(self, failing_cases):
        """When a mandatory test fails, the session should stop
        immediately.
        """
        result = failing_cases.runpytest()
        result.assert_outcomes(passed=0, failed=1)  # 2nd test shouldn't run.

    def test_session_ignore_mandatory(self, failing_cases):
        """Using --ignore-mandatory should prevent mandatory failures
        from stopping the session early.
        """
        result = failing_cases.runpytest('--ignore-mandatory')
        result.assert_outcomes(passed=0, failed=2)

    def test_message_finished(self, passing_cases):
        result = passing_cases.runpytest()
        assert 'stopping early' not in result.stdout.str()

    def test_message_stopped(self, failing_cases):
        result = failing_cases.runpytest()
        result.stdout.fnmatch_lines([
            "stopping early, mandatory 'test_first' failed",
            "use '--ignore-mandatory' to continue testing",
        ])

    def test_message_ignore_mandatory(self, failing_cases):
        result = failing_cases.runpytest('--ignore-mandatory')
        assert 'stopping early' not in result.stdout.str()


pytest_datatest.py
# -*- coding: utf-8 -*-
"""
A pytest plugin for test driven data-wrangling with datatest.

IMPORTANT: Users of Datatest should only install ``datatest``
itself, not the ``pytest_datatest`` development package.

This plugin is bundled with the datatest however, it's developed
separately. This is done for a few reasons:

1. Datatest should work as expected out-of-the-box. The plugin
   code is small enough that including it does not impact the
   user experience for non-pytest users. Offering the plugin as
   an optional dependency would add a step to the installation
   process for no real benefit.
2. Following pytest's plugin submission guidelines seems like
   good practice for long-term maintenance. But the guidelines
   aren't easily implemented for datatest as a whole. Maintaining
   the plugin in a separate repository makes it easier to follow
   the guidelines without having to retool the main datatest
   project.
3. Datatest supports more versions of Python than does Pytest
   or tox. The separate repository provides a clean separation
   between the two sets of test requirements.

Developers of the Datatest project should install both
``datatest`` and ``pytest_datatest``. When both packages
are installed, ``pytest_datatest`` is used in place of
the bundled version.
"""

import itertools
import re
import warnings

import pytest
import _pytest  # Non-public API.
from datatest import ValidationError


def _warn_import_fallback(name):
    message = 'could not import {0}; using fallback'.format(name)
    warnings.warn(message, stacklevel=2)


try:
    from _pytest.assertion.truncate import _should_truncate_item
except ImportError:
    import os

    _warn_import_fallback('_should_truncate_item')

    def _should_truncate_item(item):  # Adapted from pytest 6.1.1.
        verbose = item.config.option.verbose
        return verbose < 2 and not _running_on_ci()

    def _running_on_ci():  # Adapted from pytest 6.1.1.
        env_vars = ['CI', 'BUILD_NUMBER']
        return any(var in os.environ for var in env_vars)


try:
    from _pytest.assertion.truncate import DEFAULT_MAX_LINES
except ImportError:
    _warn_import_fallback('DEFAULT_MAX_LINES')
    DEFAULT_MAX_LINES = 8  # Adapted from pytest 6.1.1.


try:
    from _pytest.assertion.truncate import DEFAULT_MAX_CHARS
except ImportError:
    _warn_import_fallback('DEFAULT_MAX_CHARS')
    DEFAULT_MAX_CHARS = 8 * 80  # Adapted from pytest 6.1.1.


try:
    from _pytest.assertion.truncate import USAGE_MSG
except ImportError:
    _warn_import_fallback('USAGE_MSG')
    USAGE_MSG = "use '-vv' to show"  # Adapted from pytest 6.1.1.


try:
    _fail_marker = _pytest._code.code.FormattedExcinfo.fail_marker
except AttributeError:
    warnings.warn('could not reference fail_marker; using fallback')
    _fail_marker = 'E'


if __name__ == 'pytest_datatest':
    from datatest._pytest_plugin import version_info as _bundled_version_info
else:
    _bundled_version_info = (0, 0, 0)


version = '0.1.5.dev0'
version_info = (0, 1, 5)

PYTEST54 = str(pytest.__version__[:3]) == '5.4'

_idconfig_session_dict = {}  # Dictionary to store ``session`` reference.


def pytest_addoption(parser):
    """Add the '--ignore-mandatory' command line option."""
    # The following try/except block is needed because this hook
    # runs before we have a chance to turn-off the bundled plugin,
    # so this option might have already been added.
    group = parser.getgroup('Datatest')
    try:
        group.addoption(
            '--ignore-mandatory',
            action='store_true',
            help=(
                "ignore 'mandatory' marker (continues testing "
                "even when a mandatory test fails)."
            ),
        )
    except ValueError as exc:
        if 'already added' not in str(exc):
            raise


def pytest_plugin_registered(plugin, manager):
    """If running the development version, turn-off the bundled plugin."""
    development_plugin = __name__ == 'pytest_datatest'
    if development_plugin:
        manager.set_blocked(name='datatest')  # Block bundled plugin.


def pytest_configure(config):
    """Register 'mandatory' marker."""
    config.addinivalue_line(
        'markers',
        'mandatory: test is mandatory, stops session early on failure.',
    )


def pytest_collection_modifyitems(session, config, items):
    """Store ``session`` reference to use in pytest_terminal_summary()."""
    global _idconfig_session_dict
    _idconfig_session_dict[id(config)] = session


# Compile regex pattern and get fail_marker.
_diff_start_regex = re.compile(
    r'^E\s+(?:datatest.)?ValidationError:.+\d+ difference[s]?.*: [\[{]$')


def _find_validationerror_position(lines):
    """Return the index position where a ValidationError begins in the
    given list of *lines*. Return -1 if no ValidationError is found.
    """
    for position, line in enumerate(lines):
        if line.startswith(_fail_marker):
            if _diff_start_regex.search(line) is not None:
                return position  # <- EXIT!
            break  # Stop after 1st fail_marker regardless of match.
    return -1


def _formatted_lines_generator(lines, position):
    """Return a generator of formatted *lines* that contain a
    ValidationError at the given *position*.

    The resulting lines will have an unqualified class name (simply
    "ValidationError") and the fail markers ("E") will be replaced
    with spaces.
    """
    lines = iter(lines)

    # Yield lines up to the given index position without changes.
    for line in itertools.islice(lines, 0, position):
        yield line

    # Yield first failure-line, removing "datatest." and keeping fail_marker.
    fail_line = next(lines)
    yield fail_line.replace('datatest.ValidationError', 'ValidationError', 1)

    # Yield subsequent failure-lines, replacing fail_marker with spaces.
    marker_length = len(_fail_marker)
    marker_spaces = ' ' * marker_length
    for line in lines:
        if line.startswith(_fail_marker):
            yield marker_spaces + line[marker_length:]  # <- Replaces fail_marker.
        else:
            yield line
            break  # Stop checking for fail_marker after first line without a fail_marker.

    # Yield any remaining lines without changes.
    for line in lines:
        yield line


if PYTEST54:
    class ReprEntry(_pytest._code.code.ReprEntry):
        """Custom ReprEntry--USE ONLY WITH PYTEST 5.4.X VERSIONS."""
        def __init__(self, reprentry):
            self.lines = reprentry.lines
            self.reprfuncargs = reprentry.reprfuncargs
            self.reprlocals = reprentry.reprlocals
            self.reprfileloc = reprentry.reprfileloc
            self.style = reprentry.style

        def _write_entry_lines(self, tw):
            """This method is adapted from Pytest version 6.1.1."""

            if not self.lines:
                return

            fail_marker = "{0}   ".format(
                _pytest._code.code.FormattedExcinfo.fail_marker)
            indent_size = len(fail_marker)
            indents = []
            source_lines = []
            failure_lines = []
            for index, line in enumerate(self.lines):
                is_failure_line = line.startswith(fail_marker)
                if is_failure_line:
                    # from this point on all lines are considered part of the failure
                    failure_lines.extend(self.lines[index:])
                    break
                else:
                    if self.style == "value":
                        source_lines.append(line)
                    else:
                        indents.append(line[:indent_size])
                        source_lines.append(line[indent_size:])

            tw._write_source(source_lines, indents)

            # failure lines are always completely red and bold
            for line in failure_lines:
                tw.line(line, bold=True, red=True)


def _format_reprtraceback(reprtraceback):
    for reprentry in reprtraceback.reprentries:
        try:
            lines = reprentry.lines
            position = _find_validationerror_position(lines)
            if position != -1:
                lines = _formatted_lines_generator(lines, position)
                reprentry.lines = list(lines)
        except AttributeError:
            # On pytest versions 3.3 through 3.6, sessions using `xdist`
            # return `dict` instances instead of ReprEntry instances.
            lines = reprentry['lines']
            position = _find_validationerror_position(lines)
            if position != -1:
                lines = _formatted_lines_generator(lines, position)
                reprentry['lines'] = list(lines)

    if PYTEST54:
        reprtraceback.reprentries = \
            [ReprEntry(entry) for entry in reprtraceback.reprentries]


def pytest_runtest_logreport(report):
    """Hook to format the ReprEntry lines for ValidationErrors"""

    if report.when != 'call' or report.longrepr is None:
        return

    longrepr = report.longrepr
    try:
        # Try `chain` attribute (assuming ExceptionChainRepr).
        for element_tuple in longrepr.chain:
            reprtraceback = element_tuple[0]
            _format_reprtraceback(reprtraceback)
    except AttributeError:
        try:
            # Try `reprtraceback` attribute (assuming ExceptionRepr).
            _format_reprtraceback(longrepr.reprtraceback)
        except AttributeError:
            # Unknown type goes unmodified.
            pass


def _should_truncate(line_count, char_count):
    return (line_count > DEFAULT_MAX_LINES) or (char_count > DEFAULT_MAX_CHARS)


_truncation_notice = '...Full output truncated, {0}'.format(USAGE_MSG)


@pytest.hookimpl(tryfirst=True, hookwrapper=True)
def pytest_runtest_makereport(item, call):
    """Hook wrapper to replace ReprEntry instances for ValidationError
    exceptions and to handle failure of 'mandatory' tests.
    """
    if call.when == 'call':

        datafail = call.excinfo and call.excinfo.errisinstance(ValidationError)

        # Pytest-style truncation must be applied before `yield`.
        if datafail and _should_truncate_item(item):
            call.excinfo.value._should_truncate = _should_truncate
            call.excinfo.value._truncation_notice = _truncation_notice

        outcome = yield

        # If test was mandatory, session should fail immediately.
        if call.excinfo:
            try:
                mandatory = item.get_closest_marker('mandatory')
            except AttributeError:
                try:
                    mandatory = item.get_marker('mandatory')  # pytest <= 3.5
                except AttributeError:
                    mandatory = False  # in pytest <= 3.6 item can be non-Item

            if mandatory and not item.config.getoption('--ignore-mandatory'):
                shouldfail = 'mandatory {0!r} failed'.format(item.name)
                item.session.shouldfail = shouldfail

    else:
        outcome = yield  # noqa: F841 (set flake8 to ignore)


def pytest_terminal_summary(terminalreporter, exitstatus):
    """Add sections to terminal summary report when appropriate."""

    session = _idconfig_session_dict.get(id(terminalreporter.config), None)
    shouldfail = str(getattr(session, 'shouldfail', ''))
    if shouldfail.startswith('mandatory') and shouldfail.endswith('failed'):
        markup = {'yellow': True}
        terminalreporter.write_sep('_', **markup)
        terminalreporter.write(
            (
                "\n"
                "stopping early, {0}\n"
                "use '--ignore-mandatory' to continue testing\n"
                "\n"
            ).format(shouldfail),
            **markup
        )

    if _bundled_version_info > version_info:
        markup = {'yellow': True, 'bold': True}
        terminalreporter.section('NOTICE', **markup)
        terminalreporter.write(
            (
                "\n"
                "The installed version of the 'pytest_datatest' plugin "
                "is older than the bundled version included with datatest "
                "itself.\n"
                "\n"
                "Uninstall 'pytest_datatest' to automatically enable the "
                "newer version.\n"
                "\n"
            ),
            **markup
        )
